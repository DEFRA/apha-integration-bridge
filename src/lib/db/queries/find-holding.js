import Joi from 'joi'

import { query } from '../operations/query.js'

/**
 * @typedef {import('joi')} Joi
 */

/**
 * Schema for validating parameters to lookup holdings from the database
 *
 * @typedef {Object} FindHoldingSchema
 * @property {string} countyId - The ID of the county
 * @property {string} parishId - The ID of the parish
 * @property {string} holdingsId - The ID of the holdings
 */
export const FindHoldingSchema = Joi.object({
  countyId: Joi.string().regex(/^\d+$/).required().description('County ID'),
  parishId: Joi.string().regex(/^\d+$/).required().description('Parish ID'),
  holdingsId: Joi.string().regex(/^\d+$/).required().description('Holdings ID')
})

/**
 * @typedef {Record<string, Array<(value: unknown) => unknown>>} Marshallers
 * @typedef {{ sql: string; bindings: readonly unknown[]; marshallers?: Marshallers }} Query
 *
 * @param {unknown} parameters Parameters required to build the query for looking up a holding
 * @returns {Query} the query builder for looking up a holding
 */
export function findHoldingQuery(parameters) {
  /**
   * validate parameters first
   */
  const { value, error } = FindHoldingSchema.validate(parameters)

  if (error) {
    throw new Error(`Invalid parameters: ${error.message}`)
  }

  const cph = `${value.countyId}/${value.parishId}/${value.holdingsId}`

  const { sql, bindings } = query()
    .select('cph')
    .select('cph_type as cphtype')
    .distinct()
    .from('ahbrp.cph')
    .where('cph', `LIKE`, cph)
    .toSQL()

  return {
    sql,
    bindings
  }
}
