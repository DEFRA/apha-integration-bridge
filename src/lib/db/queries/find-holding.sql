SELECT
CPH.CPH cph,
'Y' cphactive, 
CPH_TYPE cphtype,
PARTY_ID cphholdercustomerid,
LOCATION_ID locationid,
ref_data_code_desc.short_Description laname,
ref_data_Code.code lanumber

FROM
AHBRP.CPH,
AHBRP.FEATURE_INVOLVEMENT FI,
AHBRP.LOCATION,
AHBRP.FEATURE_STATE FS,
AHBRP.PARTY_ROLE PR,
AHBRP.PARTY,
AHBRP.PARTY_STATE PS,
AHBRP.ref_data_Set_map,
AHBRP.ref_data_code_map,
AHBRP.ref_data_code,
AHBRP.ref_data_code ref_data_code_1,
AHBRP.ref_data_Code_desc

WHERE
CPH.CPH = FI.CPH
AND
FI.FEATURE_PK = LOCATION.FEATURE_PK
AND
FI.FEATURE_PK = FS.FEATURE_PK
AND
FI.FEATURE_INVOLVEMENT_TYPE = 'CPHHOLDERSHIP'
AND
FI.FEATURE_INVOLV_TO_DATE IS NULL
AND
FS.FEATURE_STATUS_CODE <> 'INACTIVE'
AND
FS.FEATURE_STATE_TO_DTTM IS NULL
AND
FI.PARTY_ROLE_PK = PR.PARTY_ROLE_PK
AND
PR.PARTY_PK = PARTY.PARTY_PK
AND 
PR.PARTY_ROLE_TO_DATE IS NULL
AND
PARTY.PARTY_PK = PS.PARTY_PK
AND
PS.PARTY_STATUS_CODE <> 'INACTIVE'
AND 
PS.PARTY_STATE_TO_DTTM IS NOT NULL
AND
REF_DATA_SET_MAP.REF_DATA_SET_MAP_PK = REF_DATA_CODE_MAP.REF_DATA_SET_MAP_PK
and 
REF_DATA_SET_MAP.EFFECTIVE_TO_DATE = '31/DEC/9999'
AND 
REF_DATA_SET_MAP.REF_DATA_SET_MAP_NAME = 'LOCAL_AUTHORITY_COUNTY_PARISH'
AND
REF_DATA_CODE_MAP.FROM_REF_DATA_CODE_PK = REF_DATA_CODE.REF_DATA_CODE_PK
and 
REF_DATA_CODE_MAP.EFFECTIVE_TO_DATE = '31/DEC/9999'
AND 
REF_DATA_CODE.REF_DATA_CODE_PK = REF_DATA_CODE_DESC.REF_DATA_CODE_PK
AND 
REF_DATA_CODE.EFFECTIVE_TO_DATE= '31/DEC/9999'
and 
REF_DATA_CODE_MAP.TO_REF_DATA_CODE_PK = REF_DATA_CODE_1.REF_DATA_CODE_PK
and 
REF_DATA_CODE_1.EFFECTIVE_TO_DATE = '31/DEC/9999'
AND
SUBSTR(CPH.CPH,1,6) = REF_DATA_CODE_1.CODE
AND
CPH.CPH = :cph