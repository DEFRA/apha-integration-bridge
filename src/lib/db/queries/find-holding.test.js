import { test, expect } from '@jest/globals'

import { findHoldingQuery } from './find-holding.js'

test('returns the expected query for valid parameters', () => {
  const parameters = {
    countyId: '01',
    parishId: '000',
    holdingId: '0333'
  }

  const { sql, bindings } = findHoldingQuery(parameters)

  expect(sql).toMatchSnapshot()

  expect(bindings).toEqual([
    'CPHHOLDERSHIP',
    'INACTIVE',
    'LOCAL_AUTHORITY_COUNTY_PARISH',
    '01/000/0333',
    'INACTIVE'
  ])
})

test('throws if the parameters are invalid', () => {
  expect(() => findHoldingQuery({})).toThrow(/required/i)
})
