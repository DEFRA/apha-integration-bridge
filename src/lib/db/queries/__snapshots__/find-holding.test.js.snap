// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns the expected query for valid parameters 1`] = `"select distinct cph, cph_type from ahbrp.cph where cph like ?"`;
